spark/spark-3.5/src/main/scala/org/apache/sedona/sql/datasources/geopackage/GeoPackageDataSource.scala

```diff
@@ -0,0 +1,100 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sedona.sql.datasources.geopackage
+
+import org.apache.hadoop.fs.Path
+import org.apache.sedona.sql.datasources.geopackage.model.GeoPackageOptions
+import org.apache.spark.sql.connector.catalog.Table
+import org.apache.spark.sql.execution.datasources.FileFormat
+import org.apache.spark.sql.execution.datasources.v2.FileDataSourceV2
+import org.apache.spark.sql.sources.DataSourceRegister
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+import java.util.Locale
+import scala.jdk.CollectionConverters._
+import scala.util.Try
+
+class GeoPackageDataSource extends FileDataSourceV2 with DataSourceRegister {
+
+  override def fallbackFileFormat: Class[_ <: FileFormat] = {
+    null
+  }
+
+  override protected def getTable(options: CaseInsensitiveStringMap): Table = {
+    GeoPackageTable(
+      "",
+      sparkSession,
+      options,
+      getTransformedPath(options),
+      None,
+      fallbackFileFormat,
+      getLoadOptions(options))
+  }
+
+  private def getLoadOptions(options: CaseInsensitiveStringMap): GeoPackageOptions = {
+    val path = options.get("path")
+    if (path.isEmpty) {
+      throw new IllegalArgumentException("GeoPackage path is not specified")
+    }
+
+    val showMetadata = options.getBoolean("showMetadata", false)
+    val maybeTableName = options.get("tableName")
+
+    if (!showMetadata && maybeTableName == null) {
+      throw new IllegalArgumentException("Table name is not specified")
+    }
+
+    val tableName = if (showMetadata) {
+      "gpkg_contents"
+    } else {
+      maybeTableName
+    }
+
+    GeoPackageOptions(tableName = tableName, showMetadata = showMetadata)
+  }
+
+  private def getTransformedPath(options: CaseInsensitiveStringMap): Seq[String] = {
+    val paths = getPaths(options)
+    transformPaths(paths, options)
+  }
+
+  private def transformPaths(
+      paths: Seq[String],
+      options: CaseInsensitiveStringMap): Seq[String] = {
+    val caseSensitiveMap = options.asCaseSensitiveMap.asScala.toMap
+    val hadoopConf = sparkSession.sessionState.newHadoopConfWithOptions(caseSensitiveMap)
+    paths.map { pathString =>
+      if (pathString.toLowerCase(Locale.ROOT).endsWith(".geopackage")) {
+        val path = new Path(pathString)
+        val fs = path.getFileSystem(hadoopConf)
+
+        val isDirectory = Try(fs.getFileStatus(path).isDirectory).getOrElse(false)
+        if (isDirectory) {
+          pathString
+        } else {
+          pathString.substring(0, pathString.length - 3) + "???"
+        }
```

Comment:
yeah you are right, I was sure that in this part I already have list of files in the directory, I am wondering how it would behave if I am passing list of files and if its actually needed (user specifying paths with different file formats).